### PR TITLE
Add a telamon-cli package with CLI utilities

### PR DESCRIPTION
**Pull request by Elarnon**
The telamon-cli package currently provides two executables:

 - A `cuda_search` executable which runs the search for a given kernel
   and configuration

 - A `tlcli` swiss knife executable which allows doing a bunch of handy
   and common tasks such as extracting lists of actions from event log,
   displaying eventlog statistics, replaying a list of action, etc.
   Those are scripts I had lying around that I find useful, and as such
   worth it to share.

(Yes, those should be binaries in the main `telamon` package.  They are
not because (a) it would require recompiling the main `telamon` package
each time, which is way too slow and (b) they depend on the kernels.
Yes, this should be fixed; but it requires a non negligible amount of
refactoring -- so in the meantime let's use a separate crate)

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -89,4 +89,9 @@ members = [
     "telamon-utils",
     "telamon-gen/cc_tests",
     "telamon-capi",
+    "telamon-cli",
 ]
+
+[profile.release]
+lto = false
+debug = true
```
Comment (andidr) [inline Cargo.toml]: Although trivial, I'd suggest to put this into a separate commit

File: telamon-cli/src/bin/cuda_search/main.rs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+use telamon_kernels::{linalg, Kernel};
+
+use structopt::StructOpt;
+
+/// The number of times to run the generated code to evaluate its performance.
+const NUM_CODE_RUNS: usize = 40;
+/// Search timeout in minutes.
+const TIMEOUT: u64 = 240;
+
+/// Benchamrks a kernel against a reference implementation.
```
Comment (andidr) [inline telamon-cli/src/bin/cuda_search/main.rs]: typo

File: telamon-cli/src/bin/cuda_search/main.rs [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    reference: &REF,
+    // output_dir: String,
+) where
+    K: Kernel<'a>,
+    CB: ContextBuilder<'a>,
+    REF: Reference<'a, K, Context = CB::Context>,
+{
+    config.timeout.get_or_insert(TIMEOUT);
+    // config.output_dir = output_dir;
+    //config.distance_to_best.get_or_insert(20.);
```
Comment (andidr) [inline telamon-cli/src/bin/cuda_search/main.rs]: Delete commented code

File: telamon-cli/src/bin/tlcli.rs [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+impl Bounds {
+    fn next_choice(&self, space: &SearchSpace) -> Option<Choice> {
+        if let Some(order) = &self.order {
+            explorer::choice::list(order, space).next()
+        } else {
+            explorer::choice::default_list(space).next()
+        }
+    }
+
+    /// Descend along a path in the search tree and stores the bounds encountered on the way.
```
Comment (andidr) [inline telamon-cli/src/bin/tlcli.rs]: Either Descend -> Descends or stores -> store

File: telamon-cli/src/bin/tlcli.rs [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+            }
+
+            if self.limit.map(|limit| nimpl >= limit).unwrap_or(false) {
+                break;
+            }
+        }
+
+        let stats = stats::OnlineStats::from_slice(&evalns);
+        println!(
+            "Average log10 runtime: {:.2} (± {:.2})",
```
Comment (andidr) [inline telamon-cli/src/bin/tlcli.rs]: Any particular reason why this is printed in log10?

File: telamon-cli/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,594 @@
...
+            };
+            time_cuda(|| {
+                check_cublas(cublasSgemm_v2(
+                    handle.0, op_b, op_a, n, m, k, &2., b, ldb, a, lda, &3., c, n,
+                ));
+            })
+        }
+    }
+
+    /// Reference implementation for the matrix-matrix multiplication.
```
Comment (andidr) [inline telamon-cli/src/lib.rs]: Same comment as for `matmul_reference`

File: telamon-cli/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,594 @@
...
+            time_cuda(|| {
+                check_cublas(cublasSgemmStridedBatched(
+                    handle.0, op_b, op_a, n, m, k, &2., b, ldb, stride_b, a, lda,
+                    stride_a, &3., c, n, stride_c, batch,
+                ));
+            })
+        }
+    }
+
+    /// Reference implementation for the `Gesummv` params.
```
Comment (andidr) [inline telamon-cli/src/lib.rs]: Not sure why it says "for the `Gesummv` params" instead of just "for `Gesummv`"

File: telamon-cli/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,594 @@
...
+pub enum KernelParam {
+    Axpy { n: i32 },
+    MatVec { m: i32, n: i32 },
+    Gesummv { m: i32, n: i32 },
+    Gemm { m: i32, n: i32, k: i32 },
+    BatchMM { b: i32, m: i32, n: i32, k: i32 },
+}
+
+impl KernelParam {
+    /// Build the kernel in a given context, and returns a list of candidates.
```
Comment (andidr) [inline telamon-cli/src/lib.rs]: Either "Build" -> "builds" or "returns" -> "return"

File: telamon-cli/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,594 @@
...
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ParseKernelError {
+    kind: KernelErrorKind,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum KernelErrorKind {
+    /// Value being parsed is empty.
+    ///
+    /// Among other things, this variant will be constructed when parsing an empty string.
```
Comment (andidr) [inline telamon-cli/src/lib.rs]: Not sure what the "among other things" refers to

Comment (Elarnon) [reply telamon-cli/src/lib.rs]: reused verbiage from `io::ErrorKind`.  Will remove

File: telamon-cli/src/bin/tlcli.rs (hunk 5 above)
Comment (Elarnon) [reply telamon-cli/src/bin/tlcli.rs]: Average runtime is dominated by the first few evaluations which are really bad.  Printing averages in log scale effectively computes a geometric mean which is more adapted when values have (wildly) different scales I believe.

File: telamon-cli/src/lib.rs (hunk 7 above)
Comment (Elarnon) [reply telamon-cli/src/lib.rs]: I think because it takes gesummv params as arguments.  Anyways changed it.

File: telamon-cli/src/lib.rs (hunk 6 above)
Comment (Elarnon) [reply telamon-cli/src/lib.rs]: You did not leave a comment for `matmul_reference`?

**Comment by Elarnon** [comment]
bors r=andidr

**Comment by Elarnon** [comment]
bors retry
